yang/frr-ospfd.yang

```diff
@@ -0,0 +1,957 @@
+      description
+        "Compatible with RFC 1583.";
+    }
+
+    leaf default-metric {
+      type uint32 {
+        range "0..16777214";
+      }
+      description
+        "Metric of redistributed routes.";
+    }
+
+    leaf write-multiplier {
+      type uint8 {
+        range "1..100";
+      }
+      description
+        "Maximum number of interfaces serviced per write.";
+    }
+
+    container router-info {
+      leaf as {
+        type boolean;
+        description
+          "Enable the Router Information functionality with AS flooding scope.";
+      }
+
+      leaf area {
+        type ospf-area-id;
+        description
+          "Enable the Router Information functionality with Area flooding scope.";
+      }
+    }
+  }
+
+  grouping metric-common-group {
+    leaf metric {
+      type uint32 {
+        range "0..16777214";
+      }
+      description
+        "OSPF default metric.";
+    }
+
+    leaf metric-type {
+      type enumeration {
+        enum "type-1" {
+          value 1;
+        }
+        enum "type-2" {
+          value 2;
+        }
+      }
+      description
+        "Metric type (1/2).";
+    }
+
+    leaf route-map {
+      type rmap-ref;
+      description
+        "Route map reference.";
+    }
+  }
+
+  grouping default-info-group {
+    container default-information {
+      leaf originate {
+        type boolean;
+        description
+          "Advertise a default route";
+      }
+
+      leaf always {
+        type boolean;
+        description
+          "Always advertise default route.";
+      }
+
+      uses metric-common-group;
+    }
+  }
+
+  grouping redistribute-group {
+    list redistribute {
+      key "protocol";
+      leaf protocol {
+        type frr-route-types:frr-route-types-v4;
+        description
+          "Protocol.";
+      }
+
+      uses metric-common-group;
+    }
+  }
+
+  grouping distance-group {
+    container distance {
+      leaf admin-value {
+        type uint8 {
+          range "1..255";
+        }
+        description
+          "Admin value.";
+      }
+
+      container ospf {
+        leaf external {
+          type uint8 {
+            range "1..255";
+          }
+          description
+            "Distance for external routes.";
+        }
+
+        leaf inter-area {
+          type uint8 {
+            range "1..255";
+          }
+          description
+            "Distance for inter-area routes.";
+        }
+
+        leaf intra-area {
+          type uint8 {
+            range "1..255";
+          }
+          description
+            "Distance for intra-area routes.";
+        }
+      }
+    }
+  }
+
+  grouping distribute-list-group {
+    container distribute-list {
+      list dlist {
+        key "name protocol";
+        leaf name {
+          type string;
+          description
+            "Filter networks in routing updates.";
+        }
+
+        leaf protocol {
+          type frr-route-types:frr-route-types-v4;
```

Comment:
Apparently you can specify a protocol type only when "out" is set, so maybe add a "when" substatement here to model that constraint.

Reply:
changed to out-protocol.